test/helpers/describe-contract.ts

```diff
@@ -21,6 +21,61 @@ export interface TestContext extends DeployedContracts {
 chai.use(dirtyChai);
 chai.use(solidity);
 
+export function mainnetForkTest(
+  name: string,
+  test: () => Promise<void>,
+) {
+  if (config.FORK_MAINNET) {
+    it(`Running mainnet fork test: ${name}`, test);
+  } else {
+    it.skip(`Skipping mainnet fork test: ${name}`, test);
+  }
+}
+
+export function describeContractHardhat(
+  name: string,
+  init: (ctx: TestContext) => void | Promise<void>,
+  tests: (ctx: TestContext) => void,
+): void {
+  if (!config.isHardhat() || config.FORK_MAINNET) {
+    // These tests only run on Hardhat (not including mainnet forks)
+    describe.skip(name, () => {});
+  } else {
+    // Note that the function passed into describe() should not be async.
+    describe(name, () => {
+      const ctx: TestContext = {
+        config: getDeployConfig(),
+      } as TestContext;
+
+      let preInitSnapshotId: string;
+
+      // Runs before any before() calls made within the describeContract() call.
+      before(async () => {
+        const accounts = await hre.ethers.getSigners();
+        ctx.deployer = await getDeployerSigner();
+        ctx.users = accounts.slice(1);
+
+        // Deploy contracts before taking the pre-init snapshot.
+        const deployedContracts = await getDeployedContractsOnceForTest();
+        Object.assign(
+          ctx,
+          deployedContracts,
+        );
+
+        preInitSnapshotId = await evmSnapshot();
+        await init(ctx);
+      });
+
```

Comment:
Had issues with the other `describeContract` function below since the `beforeEach` snapshot was running before every test (had issues with the `beforeEach` running even in nested `describe` blocks, reverting state that was setup beforehand). I added this function that only did one snapshot and reverted it after the tests had run.